File: src/main/java/com/dude/dms/ui/components/dialogs/RuleRunnerDialog.java
package com.dude.dms.ui.components.dialogs;

import com.dude.dms.backend.data.docs.Doc;
import com.dude.dms.backend.data.tags.Tag;
import com.dude.dms.backend.service.DocService;
import com.vaadin.flow.component.Text;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.icon.VaadinIcon;

import java.util.Map;
import java.util.Set;

public class RuleRunnerDialog extends EventDialog {

    private final Map<Doc, Set<Tag>> result;

    private final DocService docService;

    public RuleRunnerDialog(Map<Doc, Set<Tag>> result, DocService docService) {
        this.result = result;
        this.docService = docService;

        setWidth("60vw");
        setHeight("60vh");

        Text text = new Text(result.size() + " documents found");
        Button saveButton = new Button("Save", VaadinIcon.DISC.create(), event -> save());
        saveButton.setWidthFull();
        add(text, saveButton);
    }

    private void save() {
        result.keySet().forEach(doc -> {
            doc.getTags().addAll(result.get(doc));
            docService.save(doc);
            triggerEvent();
        });
        close();
    }
}

File: src/main/kotlin/com/dude/dms/ui/components/dialogs/RuleRunnerDialog.kt
package com.dude.dms.ui.components.dialogs

import com.dude.dms.backend.containers.TagContainer
import com.dude.dms.backend.data.docs.Doc
import com.dude.dms.brain.t
import com.dude.dms.utils.docService
import com.github.mvysny.karibudsl.v10.button
import com.github.mvysny.karibudsl.v10.grid
import com.github.mvysny.karibudsl.v10.onLeftClick
import com.github.mvysny.karibudsl.v10.text
import com.vaadin.flow.component.grid.Grid
import com.vaadin.flow.component.icon.VaadinIcon

class RuleRunnerDialog(private val result: Map<Doc, Set<TagContainer>>) : DmsDialog("", 60, 60) {

    init {
        val grid = grid<Pair<Doc, Set<TagContainer>>> {
            setSelectionMode(Grid.SelectionMode.MULTI)
            setItems(result.toList())
            asMultiSelect().select(result.toList())
            addColumn { it.first.guid }.setHeader("GUID")
        }
        text("${result.size} ${t("docs")}")
        button(t("save"), VaadinIcon.DISC.create()) {
            onLeftClick {
                grid.asMultiSelect().selectedItems.forEach { pair ->
                    pair.first.tags = pair.first.tags.plus(pair.second.map { it.tag })
                    docService.save(pair.first)
                }
                close()
            }
            setWidthFull()
        }
    }
}